src/main/java/com/sandeep/demoemployee/service/EmployeeService.java

```diff
@@ -0,0 +1,225 @@
+            for (Employee emp : children) {
+
+                if (emp != null)
+                    emp.setManagerId(newId);
+            }
+            children.forEach(this::addEmployee);
+        }
+    }
+
+    public boolean deleteEmployee(int id) {
+        Employee employee = this.getEmployeeById(id);
+        this.updateManager(id, employee.getManagerId());
+        employeeRepository.delete(employee);
+        return true;
+    }
+
+    public long getTotalEmployeeCount() {
+        return employeeRepository.count();
+    }
+
+    public String updateEmployee(Employee employee, Employee empOld) {
+        employee.setEmpId(empOld.getEmpId());
+        if (employee.getDesignation() != null) {
+            if (this.validateDesignation(employee, empOld.getDesignation())) {
+                empOld.setDesignation(employee.getDesignation());
+            } else
+                return "INVALID_DESIGNATION";
+        }
+        if (employee.getManagerId() != null) {
+            if (this.validateManager(empOld, this.getEmployeeById(employee.getManagerId())))
+                empOld.setManagerId(employee.getManagerId());
+            else
+                return "INVALID_PARENT";
+        }
+        if (employee.getEmpName() != null) {
+            empOld.setEmpName(employee.getEmpName());
+        }
+        employeeRepository.save(empOld);
+        return null;
+    }
+
+    private boolean validateDesignation(Employee employee, Designation designation) {
+        if (designation.getDsgnId() != employee.getDesignation().getDsgnId() && (designation.getDsgnId() == 1 || employee.getDesignation().getDsgnId() == 1))       //cannot demote the director
+        {
+            return false;
+        }
+        List<Employee> children = this.getAllByManagerId(employee.getEmpId());
+        if (!children.isEmpty())
+            for (Employee emp : children)                                                     //for children cannot be superior to manager
+            {
+                assert emp != null;
+                if (emp.getDesignation().getLvlId() <= designation.getLvlId())
+                    return false;
+            }
+        return true;
+    }
+
+    public boolean parentIsValid(Employee emp) {
+        Integer managerId = emp.getManagerId();
+        if (managerId != null && employeeRepository.existsAllByEmpIdIs(managerId)) {
+            Employee manager = this.getEmployeeById(managerId);
+            float managerLvl = manager.getDesignation().getLvlId();
+            float empLvl = emp.getDesignation().getLvlId();
+            return managerLvl < empLvl;
+        } else return emp.getDesignation().getDsgnId() == 1;
+    }
+
+    public boolean validateEntry(Employee employee) {
+        if (employee.getDesignation() != null && employee.getDesignation().getDsgnId() == 1)
+            return employee.getEmpName() == null || !(employee.getManagerId() == null || employee.getManagerId() == -1);  ///Jugaar to run the test cases
```

Comment:
line 167 and 168 should be a single if statement. That means if either of designation, employee name or managerId is null. It should not be a valid Entry.
Also employee.getEmpName()== null then ValidEntry() will return true. Which is not correct. 

This  ValidEntry need to be more simplified.